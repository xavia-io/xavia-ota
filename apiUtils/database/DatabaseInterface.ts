export interface Release {
  id: string;
  runtimeVersion: string;
  path: string;
  timestamp: string;
  commitHash: string;
  commitMessage: string;
}

export interface Tracking {
  id: string;
  releaseId: string;
  downloadTimestamp: string;
  platform: string;
}

export interface TrackingMetrics {
  platform: string;
  count: number;
}

export interface DatabaseInterface {
  createRelease(release: Omit<Release, 'id'>): Promise<Release>;
  getRelease(id: string): Promise<Release | null>;
  getReleaseByPath(path: string): Promise<Release | null>;
  listReleases(): Promise<Release[]>;
  createTracking(tracking: Omit<Tracking, 'id'>): Promise<Tracking>;
  getReleaseTrackingMetrics(releaseId: string): Promise<TrackingMetrics[]>;
  getReleaseTrackingMetricsForAllReleases(): Promise<TrackingMetrics[]>;
}
