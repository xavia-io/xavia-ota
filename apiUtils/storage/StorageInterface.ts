export interface StorageInterface {
  uploadFile(path: string, file: Buffer): Promise<string>;
  downloadFile(path: string): Promise<Buffer>;
  fileExists(path: string): Promise<boolean>;
  listFiles(directory: string): Promise<
    {
      name: string;
      updated_at: string;
      created_at: string;
      metadata: { size: number; mimetype: string };
    }[]
  >;
  listDirectories(directory: string): Promise<string[]>;
  copyFile(sourcePath: string, destinationPath: string): Promise<void>;
}
