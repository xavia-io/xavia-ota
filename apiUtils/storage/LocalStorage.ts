import fs from 'fs/promises';
import path from 'path';

import { StorageInterface } from './StorageInterface';

export class LocalStorage implements StorageInterface {
  private baseDir: string;

  constructor() {
    this.baseDir = path.join(process.cwd(), 'local-releases');
    this.ensureBaseDir();
  }

  private async ensureBaseDir() {
    try {
      await fs.access(this.baseDir);
    } catch {
      await fs.mkdir(this.baseDir, { recursive: true });
    }
  }

  private async ensureDir(dirPath: string) {
    const fullPath = path.join(this.baseDir, dirPath);
    await fs.mkdir(fullPath, { recursive: true });
  }

  async uploadFile(filePath: string, content: Buffer): Promise<string> {
    const fullPath = path.join(this.baseDir, filePath);
    await this.ensureDir(path.dirname(filePath));
    await fs.writeFile(fullPath, content);
    return filePath;
  }

  async downloadFile(filePath: string): Promise<Buffer> {
    const fullPath = path.join(this.baseDir, filePath);
    return fs.readFile(fullPath);
  }

  async fileExists(filePath: string): Promise<boolean> {
    try {
      await fs.access(path.join(this.baseDir, filePath));
      return true;
    } catch {
      return false;
    }
  }

  async listFiles(directory: string): Promise<
    {
      name: string;
      updated_at: string;
      created_at: string;
      metadata: { size: number; mimetype: string };
    }[]
  > {
    const fullPath = path.join(this.baseDir, directory);
    try {
      const files = await fs.readdir(fullPath);
      const fileStats = await Promise.all(
        files.map(async (file) => {
          const filePath = path.join(fullPath, file);
          const stats = await fs.stat(filePath);
          return {
            name: file,
            updated_at: stats.mtime.toISOString(),
            created_at: stats.birthtime.toISOString(),
            metadata: {
              size: stats.size,
              mimetype: this.getMimeType(path.extname(file)),
            },
          };
        })
      );
      return fileStats;
    } catch {
      return [];
    }
  }

  async listDirectories(directory: string): Promise<string[]> {
    const fullPath = path.join(this.baseDir, directory);
    try {
      const entries = await fs.readdir(fullPath, { withFileTypes: true });
      return entries.filter((entry) => entry.isDirectory()).map((entry) => entry.name);
    } catch {
      return [];
    }
  }

  async copyFile(sourcePath: string, destinationPath: string): Promise<void> {
    const sourceFullPath = path.join(this.baseDir, sourcePath);
    const destFullPath = path.join(this.baseDir, destinationPath);
    await this.ensureDir(path.dirname(destinationPath));
    await fs.copyFile(sourceFullPath, destFullPath);
  }

  private getMimeType(ext: string): string {
    const mimeTypes: { [key: string]: string } = {
      '.js': 'application/javascript',
      '.json': 'application/json',
      '.png': 'image/png',
      '.jpg': 'image/jpeg',
      '.jpeg': 'image/jpeg',
      '.gif': 'image/gif',
      '.zip': 'application/zip',
    };
    return mimeTypes[ext.toLowerCase()] || 'application/octet-stream';
  }
}
