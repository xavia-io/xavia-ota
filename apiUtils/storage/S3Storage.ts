import {
  S3Client,
  GetObjectCommand,
  HeadObjectCommand,
  ListObjectsV2Command,
  PutObjectCommand,
  CopyObjectCommand,
} from '@aws-sdk/client-s3';
import { StorageInterface } from './StorageInterface';

export class S3Storage implements StorageInterface {
  private client: S3Client;
  private bucketName: string;

  constructor() {
    if (!process.env.S3_ACCESS_KEY_ID || !process.env.S3_SECRET_ACCESS_KEY) {
      throw new Error('S3 credentials not configured');
    }
    if (!process.env.S3_BUCKET_NAME) {
      throw new Error('S3 bucket name not configured');
    }
    this.client = new S3Client({
      region: process.env.S3_REGION ?? 'auto',
      endpoint: process.env.S3_ENDPOINT,
      credentials: {
        accessKeyId: process.env.S3_ACCESS_KEY_ID,
        secretAccessKey: process.env.S3_SECRET_ACCESS_KEY,
      },
    });
    this.bucketName = process.env.S3_BUCKET_NAME;
  }

  async copyFile(sourcePath: string, destinationPath: string): Promise<void> {
    const copyCommand = new CopyObjectCommand({
      Bucket: this.bucketName,
      CopySource: sourcePath,
      Key: destinationPath,
    });
    await this.client.send(copyCommand);
  }

  async downloadFile(path: string): Promise<Buffer> {
    const getCommand = new GetObjectCommand({
      Bucket: this.bucketName,
      Key: path,
    });
    const response = await this.client.send(getCommand);
    const body = await response.Body?.transformToByteArray();
    if (!body) {
      throw new Error('No body found in response');
    }
    return Buffer.from(body);
  }

  async fileExists(path: string): Promise<boolean> {
    try {
      const files = await this.listDirectories(path);
      if (files.length > 0) {
        return true;
      }
    } catch {}
    try {
      const headCommand = new HeadObjectCommand({
        Bucket: this.bucketName,
        Key: path.split('/').shift(),
      });
      await this.client.send(headCommand);
      return true;
    } catch {
      return false;
    }
  }

  async listFiles(directory: string): Promise<
    {
      name: string;
      updated_at: string;
      created_at: string;
      metadata: { size: number; mimetype: string };
    }[]
  > {
    const prefix = `${directory}/`;
    const listCommand = new ListObjectsV2Command({
      Bucket: this.bucketName,
      Prefix: prefix,
    });
    const response = await this.client.send(listCommand);
    return (
      response.Contents?.map((file) => ({
        name: file.Key!.replace(prefix, ''),
        updated_at: file.LastModified?.toISOString() ?? '',
        created_at: file.LastModified?.toISOString() ?? '',
        metadata: {
          size: file.Size ?? 0,
          mimetype: this.getMimeType(file.Key?.split('.').pop() ?? 'unknown'),
        },
      })) ?? []
    );
  }

  async listDirectories(directory: string): Promise<string[]> {
    const listCommand = new ListObjectsV2Command({
      Bucket: this.bucketName,
      Prefix: directory,
      Delimiter: '/',
    });
    const response = await this.client.send(listCommand);
    return (
      response.CommonPrefixes?.map((prefix) =>
        prefix.Prefix!.replace(directory, '').replace(/\/$/, '')
      ) ?? []
    );
  }

  async uploadFile(path: string, file: Buffer): Promise<string> {
    const uploadCommand = new PutObjectCommand({
      Bucket: this.bucketName,
      Key: path,
      Body: file,
    });
    await this.client.send(uploadCommand);
    return path;
  }

  private getMimeType(ext: string): string {
    const mimeTypes: { [key: string]: string } = {
      js: 'application/javascript',
      json: 'application/json',
      png: 'image/png',
      jpg: 'image/jpeg',
      jpeg: 'image/jpeg',
      gif: 'image/gif',
      zip: 'application/zip',
    };
    return mimeTypes[ext.toLowerCase()] || 'application/octet-stream';
  }
}
