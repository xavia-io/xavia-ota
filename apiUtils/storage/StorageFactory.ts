import { LocalStorage } from './LocalStorage';
import { StorageInterface } from './StorageInterface';
import { SupabaseStorage } from './SupabaseStorage';

export class StorageFactory {
  private static instance: StorageInterface;

  static getStorage(): StorageInterface {
    if (!StorageFactory.instance) {
      const storageType = process.env.BLOB_STORAGE_TYPE;
      if (storageType === 'supabase') {
        StorageFactory.instance = new SupabaseStorage();
      } else if (storageType === 'local') {
        StorageFactory.instance = new LocalStorage();
      } else {
        throw new Error('Unsupported storage type');
      }
    }
    return StorageFactory.instance;
  }
}
