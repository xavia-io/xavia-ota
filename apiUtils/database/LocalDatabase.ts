import { Pool } from 'pg';

import { DatabaseInterface, Release, Tracking, TrackingMetrics } from './DatabaseInterface';
import { Tables } from './DatabaseFactory';

export class PostgresDatabase implements DatabaseInterface {
  private pool: Pool;

  constructor() {
    this.pool = new Pool({
      user: process.env.POSTGRES_USER,
      password: process.env.POSTGRES_PASSWORD,
      database: process.env.POSTGRES_DB,
      host: process.env.POSTGRES_HOST,
      port: parseInt(process.env.POSTGRES_PORT ?? '5432', 10),
    });
  }

  async getReleaseByPath(path: string): Promise<Release | null> {
    const query = `
      SELECT id, runtime_version as "runtimeVersion", path, timestamp, commit_hash as "commitHash"
      FROM ${Tables.RELEASES} WHERE path = $1
    `;
    const { rows } = await this.pool.query(query, [path]);
    return rows[0] || null;
  }

  async createTracking(tracking: Omit<Tracking, 'id'>): Promise<Tracking> {
    const query = `
      INSERT INTO ${Tables.RELEASES_TRACKING} (release_id, platform)
      VALUES ($1, $2)
      RETURNING id, release_id as "releaseId", download_timestamp as "downloadTimestamp", platform
    `;
    const values = [tracking.releaseId, tracking.platform];
    const { rows } = await this.pool.query(query, values);
    return rows[0];
  }

  async getReleaseTrackingMetrics(releaseId: string): Promise<TrackingMetrics[]> {
    const query = `
      SELECT platform, COUNT(*) as count
      FROM ${Tables.RELEASES_TRACKING}
      WHERE release_id = $1
      GROUP BY platform
    `;
    const { rows } = await this.pool.query(query, [releaseId]);
    return rows.map((row) => ({
      platform: row.platform,
      count: Number(row.count),
    }));
  }

  async getReleaseTrackingMetricsForAllReleases(): Promise<TrackingMetrics[]> {
    const query = `
      SELECT platform, COUNT(*) as count
      FROM ${Tables.RELEASES_TRACKING}
      GROUP BY platform
    `;
    const { rows } = await this.pool.query(query);
    return rows.map((row) => ({
      platform: row.platform,
      count: Number(row.count),
    }));
  }

  async createRelease(release: Omit<Release, 'id'>): Promise<Release> {
    const query = `
      INSERT INTO ${Tables.RELEASES} (runtime_version, path, timestamp, commit_hash, commit_message)
      VALUES ($1, $2, $3, $4, $5)
      RETURNING id, runtime_version as "runtimeVersion", path, timestamp, commit_hash as "commitHash"
    `;

    const values = [
      release.runtimeVersion,
      release.path,
      release.timestamp,
      release.commitHash,
      release.commitMessage,
    ];
    const { rows } = await this.pool.query(query, values);
    return rows[0];
  }

  async getRelease(id: string): Promise<Release | null> {
    const query = `
      SELECT id, runtime_version as "runtimeVersion", path, timestamp, commit_hash as "commitHash"
      FROM ${Tables.RELEASES} WHERE id = $1
    `;

    const { rows } = await this.pool.query(query, [id]);
    return rows[0] || null;
  }

  async listReleases(): Promise<Release[]> {
    const query = `
      SELECT id, runtime_version as "runtimeVersion", path, timestamp, commit_hash as "commitHash", commit_message as "commitMessage"
      FROM ${Tables.RELEASES}
      ORDER BY timestamp DESC
    `;

    const { rows } = await this.pool.query(query);
    return rows;
  }
}
