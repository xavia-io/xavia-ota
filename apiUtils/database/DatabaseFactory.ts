import { DatabaseInterface } from './DatabaseInterface';
import { PostgresDatabase } from './LocalDatabase';
import { SupabaseDatabase } from './SupabaseDatabase';

export enum Tables {
  RELEASES = 'releases',
  RELEASES_TRACKING = 'releases_tracking',
}

export class DatabaseFactory {
  private static instance: DatabaseInterface;

  static getDatabase(): DatabaseInterface {
    if (process.env.DB_TYPE === 'supabase') {
      DatabaseFactory.instance = new SupabaseDatabase();
    } else if (process.env.DB_TYPE === 'postgres') {
      DatabaseFactory.instance = new PostgresDatabase();
    } else {
      throw new Error('Unsupported database type');
    }
    return DatabaseFactory.instance;
  }
}
