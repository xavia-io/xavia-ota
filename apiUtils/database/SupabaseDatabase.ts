import { createClient } from '@supabase/supabase-js';

import { DatabaseInterface, Release, Tracking, TrackingMetrics } from './DatabaseInterface';
import { Tables } from './DatabaseFactory';

export class SupabaseDatabase implements DatabaseInterface {
  private supabase;

  constructor() {
    const supabaseUrl = process.env.SUPABASE_URL;
    const supabaseKey = process.env.SUPABASE_API_KEY;

    if (!supabaseUrl || !supabaseKey) {
      throw new Error('Supabase credentials not configured');
    }

    this.supabase = createClient(supabaseUrl, supabaseKey);
  }

  async getReleaseByPath(path: string): Promise<Release | null> {
    const { data, error } = await this.supabase
      .from(Tables.RELEASES)
      .select()
      .eq('path', path)
      .single();

    if (error) throw new Error(error.message);

    return data || null;
  }

  async getReleaseTrackingMetricsForAllReleases(): Promise<TrackingMetrics[]> {
    const { data: iosCount, error: iosError } = await this.supabase
      .from(Tables.RELEASES_TRACKING)
      .select('platform', { count: 'estimated', head: true })
      .eq('platform', 'ios');

    const { data: androidCount, error: androidError } = await this.supabase
      .from(Tables.RELEASES_TRACKING)
      .select('platform', { count: 'estimated', head: true })
      .eq('platform', 'android');

    if (iosError || androidError) throw new Error(iosError?.message || androidError?.message);
    return [
      {
        platform: 'ios',
        count: Number(iosCount),
      },
      {
        platform: 'android',
        count: Number(androidCount),
      },
    ];
  }
  async createTracking(tracking: Omit<Tracking, 'id'>): Promise<Tracking> {
    const { data, error } = await this.supabase
      .from(Tables.RELEASES_TRACKING)
      .insert({
        release_id: tracking.releaseId,
        platform: tracking.platform,
        download_timestamp: tracking.downloadTimestamp,
      })
      .select()
      .single();

    if (error) throw new Error(error.message);
    return data;
  }
  async getReleaseTrackingMetrics(releaseId: string): Promise<TrackingMetrics[]> {
    const { data: iosCount, error: iosError } = await this.supabase
      .from(Tables.RELEASES_TRACKING)
      .select('platform', { count: 'estimated', head: true })
      .eq('release_id', releaseId)
      .eq('platform', 'ios');

    const { data: androidCount, error: androidError } = await this.supabase
      .from(Tables.RELEASES_TRACKING)
      .select('platform', { count: 'estimated', head: true })
      .eq('release_id', releaseId)
      .eq('platform', 'android');

    if (iosError || androidError) throw new Error(iosError?.message || androidError?.message);

    return [
      {
        platform: 'ios',
        count: Number(iosCount),
      },
      {
        platform: 'android',
        count: Number(androidCount),
      },
    ];
  }

  async createRelease(release: Omit<Release, 'id'>): Promise<Release> {
    const { data, error } = await this.supabase
      .from(Tables.RELEASES)
      .insert({
        path: release.path,
        runtime_version: release.runtimeVersion,
        timestamp: release.timestamp,
        commit_hash: release.commitHash,
        commit_message: release.commitMessage,
      })
      .select()
      .single();

    if (error) throw error;
    return data;
  }

  async getRelease(id: string): Promise<Release | null> {
    const { data, error } = await this.supabase
      .from(Tables.RELEASES)
      .select()
      .eq('id', id)
      .single();

    if (error) throw error;

    return {
      id: data.id,
      path: data.path,
      runtimeVersion: data.runtime_version,
      timestamp: data.timestamp,
      commitHash: data.commit_hash,
      commitMessage: data.commit_message,
    };
  }

  async listReleases(): Promise<Release[]> {
    const { data, error } = await this.supabase
      .from(Tables.RELEASES)
      .select()
      .order('timestamp', { ascending: false });

    if (error) throw error;
    return data.map((release) => ({
      id: release.id,
      path: release.path,
      runtimeVersion: release.runtime_version,
      timestamp: release.timestamp,
      size: release.size,
      commitHash: release.commit_hash,
      commitMessage: release.commit_message,
    }));
  }
}
