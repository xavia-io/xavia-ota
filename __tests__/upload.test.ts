import formidable from 'formidable';
import { createMocks } from 'node-mocks-http';

import { DatabaseFactory } from '../apiUtils/database/DatabaseFactory';
import { StorageFactory } from '../apiUtils/storage/StorageFactory';
import uploadHandler from '../pages/api/upload';

jest.mock('../apiUtils/database/DatabaseFactory');
jest.mock('../apiUtils/storage/StorageFactory');
jest.mock('formidable');
jest.mock('fs');

describe('Upload API', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return 405 for non-POST requests', async () => {
    const { req, res } = createMocks({ method: 'GET' });
    await uploadHandler(req, res);
    expect(res._getStatusCode()).toBe(405);
    expect(JSON.parse(res._getData())).toMatchSnapshot();
  });

  it('should handle file upload successfully', async () => {
    const mockForm = {
      parse: jest.fn().mockResolvedValue([
        {
          runtimeVersion: ['1.0.0'],
          commitHash: ['abc123'],
        },
        {
          file: [{ filepath: 'test.zip' }],
        },
      ]),
    };

    (formidable as unknown as jest.Mock).mockReturnValue(mockForm);

    const mockStorage = {
      uploadFile: jest.fn().mockResolvedValue('updates/1.0.0/timestamp.zip'),
    };

    const mockDatabase = {
      createRelease: jest.fn().mockResolvedValue(true),
    };

    (StorageFactory.getStorage as jest.Mock).mockReturnValue(mockStorage);
    (DatabaseFactory.getDatabase as jest.Mock).mockReturnValue(mockDatabase);

    const { req, res } = createMocks({ method: 'POST' });
    await uploadHandler(req, res);

    expect(res._getStatusCode()).toBe(200);
    expect(JSON.parse(res._getData())).toMatchSnapshot();
    expect(mockStorage.uploadFile).toHaveBeenCalled();
    expect(mockDatabase.createRelease).toHaveBeenCalled();
  });

  it('should return 400 for missing required fields', async () => {
    const mockForm = {
      parse: jest.fn().mockResolvedValue([{}, {}]),
    };

    (formidable as unknown as jest.Mock).mockReturnValue(mockForm);

    const { req, res } = createMocks({ method: 'POST' });
    await uploadHandler(req, res);

    expect(res._getStatusCode()).toBe(400);
    expect(JSON.parse(res._getData())).toMatchSnapshot();
  });
});
