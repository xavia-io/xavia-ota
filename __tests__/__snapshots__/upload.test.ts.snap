// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload API should handle file upload successfully 1`] = `
Object {
  "path": "updates/1.0.0/timestamp.zip",
  "success": true,
}
`;

exports[`Upload API should return 400 for missing required fields 1`] = `
Object {
  "error": "Missing file, runtime version, or commit hash",
}
`;

exports[`Upload API should return 405 for non-POST requests 1`] = `
Object {
  "error": "Method not allowed",
}
`;
